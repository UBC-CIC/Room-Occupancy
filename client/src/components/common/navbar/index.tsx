import React from "react";
import { Menu, Segment, Grid, Button } from "semantic-ui-react";
import cicLogo from "../../../assets/images/cicLogo.png";
import { Link } from "react-router-dom";
import "../../../App.css";
import { useAuthenticator } from "@aws-amplify/ui-react";
import { handleSignOut } from "../../../pages/auth/Helpers";

export interface INavbarProps {}

interface NavBarPath {
  url: string;
  name: string;
}

export function Navbar(props: INavbarProps) {
  const { authStatus } = useAuthenticator((context) => [context.authStatus]);

  const navBarPaths: NavBarPath[] = [
    {
      url: "/",
      name: "Home",
    },
    {
      url: "/learn",
      name: "Learn",
    },
    {
      url: "/about",
      name: "About",
    },
    {
      url: "/analytics",
      name: "Analytics",
    },
  ];

  return (
    <nav className="navbarContainer">
      <Segment basic attached size="mini" fluid="true">
        <Grid>
          <Grid.Column width={6}>
            <img alt="logo" width={"300rem"} src={cicLogo} />
          </Grid.Column>
          <Grid.Column textAlign="center" width={5}>
            <Menu centered="true" basic="true" secondary>
              {navBarPaths.map((navPath) => {
                return (
                  <Link key={navPath.name} to={navPath.url}>
                    <Menu.Item role="menuitem">{navPath.name}</Menu.Item>
                  </Link>
                );
              })}
              {authStatus === "authenticated" && (
                <Link to="/admin_dashboard">
                  <Menu.Item role="menuitem">Admin Console</Menu.Item>
                </Link>
              )}
            </Menu>
          </Grid.Column>
          <Grid.Column textAlign="right" width={5}>
            {authStatus === "authenticated" ? (
              <Button color="blue" onClick={handleSignOut}>
                Logout
              </Button>
            ) : (
              <Link to="/login">
                <Button color="blue">Login</Button>
              </Link>
            )}
          </Grid.Column>
        </Grid>
      </Segment>
    </nav>
  );
}
