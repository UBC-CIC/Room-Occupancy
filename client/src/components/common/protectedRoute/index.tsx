import React from "react";
import { Navigate } from "react-router-dom";

export enum Roles {
  user = "user",
  admin = "admin",
}

interface Props {
  role: Roles;
  children: React.ReactNode;
}

export const ProtectedRoute = ({ role, children }: Props) => {
  if (role !== Roles.admin) {
    return <Navigate to="/" />;
  }

  return <>{children}</>;
};
