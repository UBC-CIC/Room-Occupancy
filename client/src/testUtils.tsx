import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import {
  MemoryRouter,
  MemoryRouterProps,
  BrowserRouter,
} from "react-router-dom";

// Implement with creation of providers
const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return <MemoryRouter>{children}</MemoryRouter>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

// Usage example:
// test("matches snapshot", () => {
//   // Replace 'Navbar' with your desired component and provide the props
//   testSnapshot(Navbar, { /* props for Navbar */ });
// });
const testComponentSnapshot = <P extends {}>(
  Component: React.ComponentType<P>,
  props: P,
  routerProps: MemoryRouterProps = {}
) => {
  const { asFragment } = render(
    <MemoryRouter {...routerProps}>
      <Component {...props} />
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
};

export * from "@testing-library/react";
export { customRender, testComponentSnapshot };
