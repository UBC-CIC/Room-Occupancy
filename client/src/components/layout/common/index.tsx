import React from "react";
import { Navbar } from "../../common/navbar";

type Props = {
  children: React.ReactNode;
};

export const CommonLayout = ({ children }: Props) => {
  return (
    <>
      <Navbar />
      {children}
    </>
  );
};
