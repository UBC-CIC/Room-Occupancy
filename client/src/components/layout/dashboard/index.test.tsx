import React from "react";
import { CommonLayout } from "../common";
import { DashboardLayout } from "./index";
import { customRender } from "../../../testUtils";
import { screen } from "@testing-library/react";
import "@testing-library/jest-dom";

describe("<Dashboard Layout />", () => {
  test("should match snapshot", () => {
    const { asFragment } = customRender(
      <DashboardLayout>Colin</DashboardLayout>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("should include navbar", () => {
    const { asFragment } = customRender(<CommonLayout>Colin</CommonLayout>);
    expect(screen.getByRole("navigation")).toBeInTheDocument();
    expect(screen.getByAltText("logo")).toBeInTheDocument();
  });

  test("should render children", () => {
    customRender(
      <DashboardLayout>
        <div>Test Child</div>
      </DashboardLayout>
    );
    expect(screen.getByText("Test Child")).toBeInTheDocument();
  });
});
