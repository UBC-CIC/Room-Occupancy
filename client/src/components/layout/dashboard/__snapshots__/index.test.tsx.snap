// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard Layout /> should match snapshot 1`] = `
<DocumentFragment>
  <nav
    class="navbarContainer"
  >
    <div
      class="ui mini basic attached segment"
      fluid="true"
    >
      <div
        class="ui grid"
      >
        <div
          class="six wide column"
        >
          <img
            alt="logo"
            src="test-file-stub"
            width="300rem"
          />
        </div>
        <div
          class="center aligned five wide column"
        >
          <div
            basic="true"
            centered="true"
            class="ui secondary menu"
          >
            <a
              href="/"
            >
              <div
                class="item"
                role="menuitem"
              >
                Home
              </div>
            </a>
            <a
              href="/learn"
            >
              <div
                class="item"
                role="menuitem"
              >
                Learn
              </div>
            </a>
            <a
              href="/about"
            >
              <div
                class="item"
                role="menuitem"
              >
                About
              </div>
            </a>
            <a
              href="/analytics"
            >
              <div
                class="item"
                role="menuitem"
              >
                Analytics
              </div>
            </a>
          </div>
        </div>
        <div
          class="right aligned five wide column"
        >
          <div
            class="ui buttons"
          >
            <a
              href="/login"
            >
              <button
                class="ui button"
              >
                Login
              </button>
            </a>
            <div
              class="or"
              data-text="or"
            />
            <a
              href="/signup"
            >
              <button
                class="ui blue button"
              >
                SignUp
              </button>
            </a>
          </div>
        </div>
      </div>
    </div>
  </nav>
  <div
    class="ui grid dashboardContainer"
  >
    <div
      class="two wide column sidebarContainer"
    >
      <div
        class="ui list"
        role="list"
      >
        <div
          class="item"
          role="listitem"
        >
          <div
            class="header"
          >
            ANALYTICS
          </div>
          <div
            class="content"
          >
            <div
              class="ui list"
              role="list"
            >
              <div
                class="item"
                role="listitem"
              >
                <i
                  aria-hidden="true"
                  class="grey chart area icon"
                />
                <a
                  class="content"
                >
                  Analytics
                </a>
              </div>
              <div
                class="item"
                role="listitem"
              >
                <i
                  aria-hidden="true"
                  class="grey camera icon"
                />
                <a
                  class="content"
                >
                  Camera
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          class="item"
          role="listitem"
        >
          <div
            class="header"
          >
            LOCATIONS
          </div>
          <div
            class="content"
          >
            <div
              class="ui list"
              role="list"
            >
              <div
                class="item"
                role="listitem"
              >
                <i
                  aria-hidden="true"
                  class="grey location arrow icon"
                />
                <a
                  class="content"
                >
                  Zone 1
                </a>
              </div>
              <div
                class="item"
                role="listitem"
              >
                <i
                  aria-hidden="true"
                  class="grey location arrow icon"
                />
                <a
                  class="content"
                >
                  Zone 2
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="fourteen wide column"
    >
      Colin
    </div>
  </div>
</DocumentFragment>
`;
